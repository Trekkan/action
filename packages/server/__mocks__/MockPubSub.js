import dataFieldsToSerialize from '../__tests__/utils/dataFieldsToSerialize'
import * as getPubSub from '../utils/getPubSub'

const fieldsToSerialize = {
  invitation: [...dataFieldsToSerialize],
  newAuthToken: ['channelId', 'message.data.tms'],
  notification: [...dataFieldsToSerialize],
  organization: [...dataFieldsToSerialize],
  task: [...dataFieldsToSerialize],
  team: [...dataFieldsToSerialize],
  teamMember: [...dataFieldsToSerialize]
}

const untestablePayloads = ['message.data.removedRequestNotifications']

// const getPath = (str, obj) => str.split('.').reduce((o, i) => o[i], obj);
const replacePath = (str, docArr, newVal) => {
  docArr.forEach((obj) => {
    const path = str.split('.')
    let currentObj = obj
    for (let ii = 0; ii < path.length - 1; ii++) {
      const next = path[ii]
      currentObj = currentObj[next]
      if (!currentObj) return
    }
    const lastVal = path[path.length - 1]
    if (Array.isArray(currentObj[lastVal])) {
      currentObj[lastVal] = newVal
    }
  })
}

export default class MockPubSub {
  constructor () {
    getPubSub.default = () => this
    this.db = {}
  }

  __serialize (dynamicSerializer) {
    const channels = Object.keys(this.db).sort()
    const snapshot = {}
    for (let i = 0; i < channels.length; i++) {
      const channel = channels[i]
      const doc = this.db[channel]
      const channelFields = fieldsToSerialize[channel]
      if (!channelFields) {
        throw new Error(`BAD MOCK: No fieldsToSerialize for pubsub channel ${channel}`)
      }

      // some payloads can't be tested because they are an array that can't be sorted
      // for those, we can skip it here and use the arrayContaining test
      untestablePayloads.forEach((untestable) => {
        replacePath(untestable, doc, 'UNTESTED')
      })
      snapshot[channel] = dynamicSerializer.toStatic(doc, channelFields)
    }
    return snapshot
  }

  publish (channel, message) {
    const [channelName, channelId] = channel.split('.')
    this.db[channelName] = this.db[channelName] || []
    this.db[channelName].push({channelId, message})
  }
}
