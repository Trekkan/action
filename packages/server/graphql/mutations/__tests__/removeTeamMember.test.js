/* eslint-env jest */
import DynamicSerializer from 'dynamic-serializer'
import MockDate from 'mockdate'
import MockPubSub from '../../../__mocks__/MockPubSub'
import makeDataLoader from '../../../__tests__/setup/makeDataLoader'
import mockAuthToken from '../../../__tests__/setup/mockAuthToken'
import MockDB from '../../../__tests__/setup/MockDB'
import {__now} from '../../../__tests__/setup/mockTimes'
import fetchAndSerialize from '../../../__tests__/utils/fetchAndSerialize'
import getRethink from '../../../database/rethinkDriver'
import removeTeamMember from '../removeTeamMember'

MockDate.set(__now)
console.error = jest.fn()

describe('removeTeamMember', () => {
  test('promotes another member if the person removed was the lead', async () => {})

  test('removes the teamMember, reassigns active tasks to the lead', async () => {
    // SETUP
    const r = await getRethink()
    const mockPubSub = new MockPubSub()
    const mockDB = new MockDB()
    const dynamicSerializer = new DynamicSerializer()
    await mockDB.init().newTask()
    const userToBoot = mockDB.db.user[7]
    const authToken = mockAuthToken(mockDB.db.user[0])
    const dataLoader = makeDataLoader(authToken)
    const teamId = mockDB.db.team[0].id
    // TEST
    const teamMemberId = mockDB.db.teamMember[7].id
    await removeTeamMember.resolve(undefined, {teamMemberId}, {authToken, dataLoader})

    const db = await fetchAndSerialize(
      {
        teamMember: r
          .table('TeamMember')
          .getAll(teamId, {index: 'teamId'})
          .orderBy('checkInOrder'),
        task: r.table('Task').getAll(teamId, {index: 'teamId'}),
        user: r.table('User').get(userToBoot.id)
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot()
  })

  test('return error if the caller is not self or team lead', async () => {
    // reassigns their tasks

    // deactivates providers

    // inserts notification in table if kickout

    // sends notification via pubsub

    // remove github repos

    // archive githug tasks attached to those repos
    // SETUP
    const mockDB = new MockDB()
    await mockDB.init().newTask()
    const authToken = mockAuthToken(mockDB.db.user[1])
    const dataLoader = makeDataLoader(authToken)
    // TEST
    const teamMemberId = mockDB.db.teamMember[7].id
    const res = await removeTeamMember.resolve(undefined, {teamMemberId}, {authToken, dataLoader})
    expect(res).toEqual(expect.objectContaining({error: expect.any(Object)}))
  })
})
