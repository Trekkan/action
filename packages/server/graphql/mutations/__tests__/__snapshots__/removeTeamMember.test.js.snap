// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeTeamMember removes the teamMember, reassigns active tasks to the lead 1`] = `
Object {
  "task": Array [
    Object {
      "assigneeId": 0,
      "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Task[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
      "createdAt": 2017-01-01T06:59:59.999Z,
      "createdBy": 24,
      "id": 25,
      "sortOrder": 0,
      "status": "active",
      "tags": Array [],
      "teamId": 16,
      "updatedAt": 2017-01-01T07:00:00.000Z,
      "userId": 17,
    },
  ],
  "teamMember": Array [
    Object {
      "checkInOrder": 8,
      "email": "steve.rogers@avengers.org",
      "id": 0,
      "isFacilitator": false,
      "isLead": true,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Captain America",
      "teamId": 16,
      "userId": 17,
    },
    Object {
      "checkInOrder": 9,
      "email": "vision@avengers.org",
      "id": 1,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Vision",
      "teamId": 16,
      "userId": 18,
    },
    Object {
      "checkInOrder": 10,
      "email": "james.rhodes@avengers.org",
      "id": 2,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "War Machine",
      "teamId": 16,
      "userId": 19,
    },
    Object {
      "checkInOrder": 11,
      "email": "tony.stark@avengers.org",
      "id": 3,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Iron Man",
      "teamId": 16,
      "userId": 20,
    },
    Object {
      "checkInOrder": 12,
      "email": "bruce.banner@avengers.org",
      "id": 4,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Hulk",
      "teamId": 16,
      "userId": 21,
    },
    Object {
      "checkInOrder": 13,
      "email": "natasha.romanova@avengers.org",
      "id": 5,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Black Widow",
      "teamId": 16,
      "userId": 22,
    },
    Object {
      "checkInOrder": 14,
      "email": "kyle.richmond@avengers.org",
      "id": 6,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Nighthawk",
      "teamId": 16,
      "userId": 23,
    },
    Object {
      "checkInOrder": 15,
      "email": "thor@avengers.org",
      "id": 7,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": false,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Thor",
      "teamId": 16,
      "updatedAt": 2017-01-01T08:00:00.000Z,
      "userId": 24,
    },
  ],
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "thor@avengers.org",
    "emailVerified": false,
    "id": 24,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "Thor Odinson",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Thor",
    "tms": Array [],
    "updatedAt": 2017-01-01T07:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 26,
        "role": null,
      },
      Object {
        "id": 27,
        "role": "billingLeader",
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;

exports[`removeTeamMember removes the teamMember, reassigns active tasks to the lead 2`] = `
Object {
  "newAuthToken": Array [
    Object {
      "channelId": 24,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
  ],
  "task": Array [
    Object {
      "channelId": 28,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 28,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 28,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 28,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 28,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 28,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 28,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
  ],
  "team": Array [
    Object {
      "channelId": 24,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
  ],
  "teamMember": Array [
    Object {
      "channelId": 16,
      "message": Object {
        "data": Object {
          "notificationId": 29,
          "removedNotifications": Array [],
          "taskIds": Array [
            37,
          ],
          "teamId": 16,
          "teamMemberId": 7,
          "type": "RemoveTeamMemberPayload",
          "userId": 24,
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
  ],
}
`;
